src/lib/crypto/kimchi_backend/gadgets/rot.ml

```diff
@@ -0,0 +1,164 @@
+open Core_kernel
+
+open Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint
+
+module Bignum_bigint = Snarky_backendless.Backend_extended.Bignum_bigint
+
+(* ROT *)
+
+(* Side of rotation *)
+type rot_mode = Left | Right
+
+(* 64-bit rotation of rot_bits to the `mode` side *)
+let rotate (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (word : Circuit.Field.t) (rot_bits : int) (mode : rot_mode) :
+    Circuit.Field.t =
+  let open Circuit in
+  (* Check that the rotation bits is smaller than 64 *)
+  assert (rot_bits < 64) ;
+  (* Check that the rotation bits is non-negative *)
+  assert (rot_bits >= 0) ;
+
+  (* Compute actual length depending on whether the rotation mode is Left or Right *)
+  let rot_bits = match mode with Left -> rot_bits | Right -> 64 - rot_bits in
+
+  (* Auxiliary Bignum_bigint values *)
+  let big_2_pow_64 = Bignum_bigint.(pow (of_int 2) (of_int 64)) in
+  let big_2_pow_rot = Bignum_bigint.(pow (of_int 2) (of_int rot_bits)) in
+
+  (* Compute the rotated word *)
+  let values =
+    exists (Typ.array ~length:4 Field.typ) ~compute:(fun () ->
+        (* Assert that word is at most 64 bits*)
+        let word_big =
+          Common.(
+            field_to_bignum_bigint
+              (module Circuit)
+              (cvar_field_to_field_as_prover (module Circuit) word))
+        in
+        assert (Bignum_bigint.(word_big < big_2_pow_64)) ;
+
+        (* Obtain rotated output, excess, and shifted for the equation
+             word * 2^rot = excess * 2^64 + shifted *)
+        let excess_big, shifted_big =
+          Common.bignum_bigint_div_rem
+            Bignum_bigint.(word_big * big_2_pow_rot)
+            big_2_pow_64
+        in
+
+        (* Compute rotated value as
+           rotated = excess + shifted *)
+        let rotated_big = Bignum_bigint.(shifted_big + excess_big) in
+
+        (* Compute bound that is the right input of FFAdd equation *)
+        let bound_big =
+          Bignum_bigint.(excess_big + big_2_pow_64 - big_2_pow_rot)
```

Comment:
Double check this is correct.  Isn't it `2^64 - 2^rot` as it is in the rust version?

Reply:
No because in the rust version the bound operation is performed later in the witness code